### PR TITLE
Cctri 161

### PR DESCRIPTION
**Pull request by cawabanga2811**
Complete campaign API with success tests. Test that failed will be done on the next day. 

Few questions I want to highlight here:
- The naming of API's. I try to stay close to official documentation, maybe someone can create a better naming.
- test assertion
- many if's

### CODE REVIEW

File: threatresponse/client.py [hunk 1]
```diff
@@ -22,6 +23,7 @@ def __init__(self, client_id, client_password, **options):
 
         self._inspect = InspectAPI(request)
         self._enrich = EnrichAPI(request)
+        self._campaign = CampaignAPI(request)
```
Comment (AlinaBlahun) [inline threatresponse/client.py]: As I understand, we should create RelativeRequest(request, url_for(region, 'private_intel')) for this bunch of endpoints.

File: threatresponse/api/campaign.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                array_for_url(search_after) + '?' +
+                urlencode(query))
+        else:
+            response = self._request.get(
+                '/ctia/campaign/search' + '?' +
+                urlencode(query))
+        response.raise_for_status()
+        return response.json()
+
+    @route('campaign')
```
Comment (AlinaBlahun) [inline threatresponse/api/campaign.py]: Shouldn't it be campaign.post campaign.delete campaign.put  and so on?

File: threatresponse/client.py (hunk 1 above)
Comment (cawabanga2811) [reply threatresponse/client.py]: + 
Sure, will add this.

File: threatresponse/api/campaign.py (hunk 2 above)
Comment (cawabanga2811) [reply threatresponse/api/campaign.py]: As I see, we just define this in chat so I will redo.

File: tests/unit/api/test_coa.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        id_=12,
+        url='/ctia/coa/12',
+        payload={'ham': 'egg'},
+    )
+
+
+def assert_succeeds_with_get(invoke, url, id_=None, **query):
+    response, request, api = response_request_and_api()
+    request.get.return_value = response
+    if id_ and query:
```
Comment (konstantin-ogulchansky) [inline tests/unit/api/test_coa.py]: We could probably rewrite it like this:
``` python
invoke(api, *params(id_, query))
```
Where `params` is defined as follows:
``` python
def params(*args):
    return [arg for arg in args if arg is not None]
```

File: tests/unit/api/test_data_table.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    if payload is not None:
+        request.post.assert_called_once_with(url, json=payload)
+    else:
+        request.post.assert_called_once_with(url)
+
+    response.json.assert_called_once_with()
+
+
+def assert_succeeds_with_delete(invoke, url, id_):
```
Comment (konstantin-ogulchansky) [inline tests/unit/api/test_data_table.py]: Looks like there is a lot of duplicated code. Could we reuse some of these functions between other tests?

File: threatresponse/api/campaign.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+from .routing import Router
+from .base import API
+
+
+class CampaignAPI(API):
```
Comment (konstantin-ogulchansky) [inline threatresponse/api/campaign.py]: It seems like implementations of `CampaignAPI`, `COAAPI` and `DataTableAPI` are identical. What do you think about having some `EntityAPI`, that will be used like this:
``` python
class CampaignAPI(EntityAPI):
    """ https://private.intel.amp.cisco.com/index.html#!/Campaign/ """

    def __init__(self):
        super(CampaignAPI, self).__init__('/ctia/campaign')
```
The implementation of `EntityAPI` will contain `get`, `post`, `external_id` and other required methods.

Comment (konstantin-ogulchansky) [reply threatresponse/api/campaign.py]: I have implemented `EntityAPI` within CCTRI-170, so you can take a look and tell me what you think. :)

Comment (cawabanga2811) [reply threatresponse/api/campaign.py]: Oh, I will, thx!

File: tests/unit/api/test_data_table.py (hunk 4 above)
Comment (cawabanga2811) [reply tests/unit/api/test_data_table.py]: Sure, we just need to move them to some assertions.py or smth.

File: threatresponse/api/intel.py [hunk 6]
```diff
@@ -7,10 +7,32 @@ class IntelAPI(API):
...
         super(IntelAPI, self).__init__(request)
 
-        self._actor = EntityAPI(request, '/ctia/actor/')
+        self._actor = EntityAPI(request, '/ctia/actor')
         self._actor.__doc__ = \
             "https://private.intel.amp.cisco.com/index.html#!/Actor/"
+        self._campaign = EntityAPI(request, '/ctia/campaign')
+        self._campaign.__doc__ = \
+            "https://private.intel.amp.cisco.com/index.html#!/Campaign/"
+        self._coa = EntityAPI(request, '/ctia/coa')
```
Comment (konstantin-ogulchansky) [inline threatresponse/api/intel.py]: Let's put some whitespaces between `_actor`, `_campaign`, `_coa` and `_data_table`.

File: tests/unit/api/test_campaign.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
+from mock import MagicMock
+
+from threatresponse.api import IntelAPI
+
+
+def test_campaign_by_id_succeeds():
```
Comment (konstantin-ogulchansky) [inline tests/unit/api/test_campaign.py]: I think that we can just test `EntityAPI` instead of testing `campaing`, `coa` and `data_table`.

Comment (cawabanga2811) [reply tests/unit/api/test_campaign.py]: true. will be removed later.

**Pull request merged**